### PR TITLE
Factorize external input code

### PR DESCRIPTION
**Pull request by toots**

### CODE REVIEW

File: libs/video.liq [hunk 1]
```diff
@@ -37,5 +37,7 @@ end
 # ffmpeg's test source video (useful for testing and debugging).
 # @category Source / Video Processing
 def video.external.testsrc()
-  input.external.avi("ffmpeg -f lavfi -i testsrc -f avi -vf format=rgb24 -vcodec rawvideo -r #{video.frame.rate()} -s #{video.frame.width()}x#{video.frame.height()} pipe:1")
+  audio_rate = get(default=-1,"frame.audio.samplerate")
```
Comment (smimram) [inline libs/video.liq]: I just added `audio.samplerate()`.

**Comment by smimram** [comment]
Testing with
```
s = input.external.ffmpeg("-i test.ogv")
s = mksafe(s)
out(s)
output.sdl(drop_audio(s))
```
It seems to loop on the first buffer (= I hear the first second looped over and over). It is not doing this on master.

**Comment by toots** [comment]
> Testing with
> 
> ```
> s = input.external.ffmpeg("-i test.ogv")
> s = mksafe(s)
> out(s)
> output.sdl(drop_audio(s))
> ```
> 
> It seems to loop on the first buffer (= I hear the first second looped over and over). It is not doing this on master.

Should be fixed now.. Programming logic issue sorry folks!

**Comment by smimram** [comment]
I can confirm that it seems to work now!

**Comment by dhannyz** [comment]
I'm getting some strange behaviour when trying to stream with OBS, The connection seems to be frequently closed/reopened, then when doing a stopping OBS, observing the input.external.avi restart in liq, and then restarting the OBS stream, I'm then met with this panic

```
2019/06/24 14:55:04 [live_video:4] [avi @ 0x556acba92e00] rgb24 rawvideo cannot be written to avi, output file will be unreadable

2019/06/24 14:53:17 [live_video:3] Error while running process: Failure("Only one audio track is supported for now.")
Raised at file "tools/process_handler.ml", line 223, characters 28-47
Called from file "tools/process_handler.ml" (inlined), line 226, characters 8-68
Called from file "tools/process_handler.ml", line 241, characters 45-84
Called from file "list.ml", line 121, characters 24-34
Called from file "tools/process_handler.ml", line 240, characters 28-321

2019/06/24 14:55:04 [threads:2] Queue generic queue #3 crashed with exception Failure("Only one audio track is supported for now.")
Raised at file "tools/process_handler.ml", line 287, characters 34-41
Called from file "duppy.ml", line 151, characters 61-77
Called from file "duppy.ml", line 279, characters 14-23
Called from file "duppy.ml", line 313, characters 7-24
Called from file "duppy.ml", line 356, characters 5-9
Called from file "tools/tutils.ml", line 214, characters 19-55

2019/06/24 14:55:04 [threads:1] PANIC: Liquidsoap has crashed, exiting.,
```

**Comment by smimram** [comment]
@dhannyz can you share some (preferably minimized) script to reproduce the issue?

**Comment by smimram** [comment]
The "one audio track" failure should be fixed by bb0cd0434006da0867dec950756e79150b59e72a.

**Comment by smimram** [comment]
@toots it seems that the buffer is not empty when the process is restarted. Because of this the second `read_header` often does not read a valid header (I say often: some thread must add some data after the process is stopped).

**Comment by toots** [comment]
> @toots it seems that the buffer is not empty when the process is restarted. Because of this the second `read_header` often does not read a valid header (I say often: some thread must add some data after the process is stopped).

Try now?

**Comment by smimram** [comment]
> > @toots it seems that the buffer is not empty when the process is restarted. Because of this the second `read_header` often does not read a valid header (I say often: some thread must add some data after the process is stopped).
> 
> Try now?

Seems all good to me!

File: libs/video.liq [hunk 2]
```diff
@@ -39,15 +39,15 @@ end
 # @param input ffmpeg options specifying the input
 def input.external.ffmpeg(~id="", ~show_command=false, input)
```
Comment (toots) [inline libs/video.liq]: Please wrap up the other parameters from `output.external.avi`, in particular `fallible`..

**Comment by dhannyz** [comment]
So the restarting issue looks to be solved.
Now my issue is the stream speed/connectivity, It seems the way data is read out of ffmpeg is a bit different?
As i have no connectivity problem on master, holding a solid speed and reading in 'realtime' however on this branch the same ffmpeg command is only reading in at a low speed, barely 0.5x, and the fps looks wonky as well.
This is causing the input buffer to be depleted every few seconds and the source going up & down.

**Comment by toots** [comment]
I can confirm the slow read. Not sure yet what's going on compared to `master`, will have a look at it. Thanks for your patience.

**Comment by toots** [comment]
Ok guys, we're looking good now! We were doing too many data copies, I've optimized all that now!

**Comment by smimram** [comment]
I can confirm it's much better.

**Comment by dhannyz** [comment]
I can confirm this is now working nicely :D
I can now stream via OBS smoothly to the remote liquidsoap server, which forwards the video stream to youtube. Disconnect and reconnect working nicely.

**Comment by toots** [comment]
Great! Merging this one then. We're getting close to release time! \o/

**Pull request merged**